Add event listener doc (#212)
enhance-dev/enhance.dev: app/docs/md/patterns/event-listeners.md

---
title: Event Listeners
---

An event is a signal that something has happened on your page. The browser notifies you so you can react to them.

## What are events?

From [MDN](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#what_is_an_event)

> Events are things that happen in the system you are programming — the system produces (or "fires") a signal of some kind when an event occurs, and provides a mechanism by which an action can be automatically taken (that is, some code running) when the event occurs. Events are fired inside the browser window, and tend to be attached to a specific item that resides in it. This might be a single element, a set of elements, the HTML document loaded in the current tab, or the entire browser window. There are many different types of events that can occur.

For example:

- A user submits a form
- A user clicks on a button
- A web page finishes loading
- etc.

**[Learn more about events](https://developer.mozilla.org/en-US/docs/Web/Events)**

## Adding a click event listener

> Suppose we have a simple web component called `custom-button` that we want to add a click event listener to.

Reviewer: Thinking this should maybe just be a vanilla WC rather than extending `CustomElement`, since we're not really gaining anything from it (except for style scoping, which we likely don't need to show here). Since events are so basic I feel like adding an abstraction off the bat may confuse some folks 🤔

Reviewer 2: I could go either way on this one. The user could be turned off by the amount of boilerplate needed if we show the stock web component.

Reviewer: Hmm, yeah since this is a client side component I guess so — an SSR element would be simpler but then we'd have to explain that whole divide. Fair!

Author: I worry about doing a vanilla WC here as we'd lose the ability to SSR it and it wouldn't be consistent with the rest of the site.

Reviewer: All good! Will close this one.

<doc-code filename="app/components/custom-button.mjs">

```javascript
import CustomElement from '@enhance/custom-element'

export default class CustomButton extends CustomElement {
  render ({ html, state }) {
    return html`
        <style>
            button {
                border: 1px solid white;
                padding: 2px;
            }
        </style>
        <button>Click me!</button>
    `
  }
}

customElements.define('custom-button', CustomButton)
```

</doc-code>

> As a best practice, event listeners for web components should be added inside the `connectedCallback` function, since this method runs when the element is added to the DOM.

Reviewer: IIRC, @kristoferjoseph just ran into an issue w/ DSD where event listeners need to be registered in the constructor instead? May need to clarify here in case.

Reviewer 2: With DSD and SSR a user does have access to the entire template generated DOM but `connectedCallback` is the correct approach that to document since it works for both but only DSD and SSR support adding event listeners in the constructor.

The main issue being that even if the user is able to attach listeners in the constructor after initial render they may try to use the same class client side without SSR or DSD and be confused as to why it doesn’t work.

Reviewer: Ahh, so I had it backwards? Coooool 😄

Inside our `connectedCallback` method, we'll use the `querySelector` method on the custom element, to get the button inside our component. Then we will add a `click` listener to the button.

<doc-code filename="app/components/custom-button.mjs">

```javascript
import CustomElement from '@enhance/custom-element'

export default class CustomButton extends CustomElement {
  connectedCallback () {
    this.button = this.querySelector('button')
    this.button.addEventListener('click', function (event) {
      alert('I was clicked')
    })
  }

  render ({ html, state }) {
    return html`
        <style>
            button {
                border: 1px solid white;
                padding: 2px;
            }
        </style>
        <button>Click me!</button>
    `
  }
}

customElements.define('custom-button', CustomButton)
```

</doc-code>

> That works great but if we truly do want to follow best practices we should clean up our event listeners when our component is removed from the DOM. We'll remove that event listener in the `disconnectedCallback` method. In order to do that we'll need to refactor our component slightly.

Reviewer: ```suggestion
That works great, but if we want to follow best practices, we should clean up our event listeners when our component is removed from the DOM. We'll remove that event listener in the `disconnectedCallback` method. In order to do that, we'll need to refactor our component slightly.
```

1. Extract our click handler to a method named `clickHander`.
1. Save a reference to the `button` element in the `constructor`.
1. Bind `this` to the callback function.
1. Add the event listener in `connectedCallback`.
1. Remove the event listener in `disconnectedCallback`.

<doc-code filename="app/components/custom-button.mjs">

```javascript
import CustomElement from '@enhance/custom-element'

export default class CustomButton extends CustomElement {
  handleClick (event) {
    alert('I was clicked')

>   }

Reviewer: This is SUPER personal preference I reckon, but I'd move `handleClick` to follow after the constructor. I feel like if I don't see the constructor first in a WC it's a bit odd. Feel free to ignore!

  constructor () {
    super()
    this.button = this.querySelector('button')
    this.handleClick = this.handleClick.bind(this)
  }

  connectedCallback () {
    this.button.addEventListener('click', this.handleClick)
  }

  disconnectedCallback () {
    this.button.removeEventListener('click', this.handleClick)
  }

  render ({ html, state }) {
    return html`
        <style>
            button {
                border: 1px solid white;
                padding: 2px;
            }
        </style>
        <button>Click me!</button>
    `
  }
}

customElements.define('custom-button', CustomButton)
```

</doc-code>

<doc-callout level="info" mark="ℹ️">

> If the bind syntax offends you, then you can always use a public class field and an arrow function instead.

Reviewer: ```suggestion
If [the `bind` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind) offends you, then you can always use [a public class field and an arrow function](https://www.aleksandrhovhannisyan.com/blog/javascript-fields-vs-methods/#alternatives-for-this-binding) instead.
```

Not sure if these are the best links to use, but I reckon these terms could be new to a lot of folks, so having some kind of reference for these is likely helpful!

```javascript
  handleClick = (event) => {
    alert('I was clicked')
  }
```

</doc-callout>

## Reducing Boilerplate

> Many other web components provide a way of reducing the amount of boilerplate code one needs to write. Enhance provides the `@enhance/element` package which builds upon the `CustomElement` and `EventHandlerMixin` classes while providing a more succinct way of writing Enhance Components.

Reviewer: ```suggestion
Many web component libraries provide a way of reducing boilerplate when authoring web components. Enhance provides the `@enhance/element` package, which builds upon the [`CustomElement`](https://github.com/enhance-dev/enhance-custom-element) and [`EventHandlerMixin`](https://github.com/enhance-dev/enhance-event-handler-mixin) classes while providing a more succinct way of writing Enhance Components.
```
